### PR TITLE
WIP: Add distinct for BSI fields at the fragment level

### LINKED ISSUE #2018: get DISTINCT values for an int (BSI) field
### Description
A question was posed by @brucenan here: https://github.com/travisturner/bit-slice-index/issues/1 about getting the set of distinct values for a BSI field.

We can currently get the value for a single column, but I don't believe we have anything that scans the BSI fragment data and collects multiple values.

I wonder if a solution might have multiple row scanners (one for each BSI row), which synchronously scan across the columns to determine each column's value. That value can then be added to a set, which would be the fragment's return value for the distinct query. The aggregator would have to merge the sets from all fragments. Each fragment would have to scan up to 2^20 columns (at standard shardWidth).

### Success criteria (What criteria will consider this ticket closeable?)
Once can get the set of distinct values from an int field.

### PR DESCRIPTION
**Pull request by travisturner**
## Overview

This is a first stab at implementing DISTINCT on a BSI field.

This solution uses an array of channels to read each bit value by column. An alternative may be to maintain an array of sums of size `consider`.

I plan to add some benchmarks next, and then I'll implement that alternative method to see how it differs.

TODO:
- [ ] Bubble `distinct` up to executor
- [ ] Add Distinct() to PQL

Fixes #2018

## Pull request checklist

- [x] I have read the [contributing guide](https://github.com/pilosa/pilosa/blob/master/CONTRIBUTING.md).
- [x] I have agreed to the [Contributor License Agreement](https://cla-assistant.io/pilosa/pilosa).
- [ ] I have updated the [documentation](https://github.com/pilosa/pilosa/tree/master/docs).
- [x] I have resolved any merge conflicts.
- [x] I have included tests that cover my changes.
- [x] All new and existing tests pass.
- [x] Make sure PR title conforms to convention in CHANGELOG.md.
- [x] Add appropriate changelog label to PR (if applicable).

## Code review checklist
This is the checklist that the reviewer will follow while reviewing your pull request. You do not need to do anything with this checklist, but be aware of what the reviewer will be looking for.

- [ ] Ensure that any changes to external docs have been included in this pull request.
- [ ] If the changes require that minor/major versions need to be updated, tag the PR appropriately.
- [ ] Ensure the new code is [properly commented](https://github.com/golang/go/wiki/CodeReviewComments#doc-comments) and follows [Idiomatic Go](https://dmitri.shuralyov.com/idiomatic-go).
- [ ] Check that tests have been written and that they cover the new functionality.
- [ ] Run tests and ensure they pass.
- [ ] Build and run the code, performing any applicable integration testing.
- [ ] Make sure PR title conforms to convention in CHANGELOG.md.
- [ ] Make sure PR is tagged with appropriate changelog label.

### CODE REVIEW

File: fragment.go [hunk 1]
```diff
@@ -1104,6 +1104,108 @@ func (f *fragment) importSetValue(columnID uint64, bitDepth uint, value int64, c
...
+		go func(rowID uint) {
+			rowToChan(consider, f.row(uint64(bsiOffsetBit+rowID)), chans[rowID])
+		}(i)
+	}
+	go func() {
+		rowToChan(consider, nrow, signCh)
+	}()
+
+	// set is a map used to maintain a distinct set of return values.
+	type void struct{}
```
Comment (tgruben) [inline fragment.go]: nice

**Comment by travisturner** [comment]
closing in favor of #2053

File: fragment.go [hunk 2]
```diff
@@ -1104,6 +1104,107 @@ func (f *fragment) importSetValue(columnID uint64, bitDepth uint, value int64, c
 	return changed, nil
 }
 
+// distinct returns a list of distinct values in the fragment.
+// A bitmap can be passed in to optionally filter the computed columns.
+func (f *fragment) distinct(filter *Row, bitDepth uint) ([]int64, error) {
```
Comment (brucenan) [inline fragment.go]: How about the performance about this function? For 10,000,000 columns, how long does it spend to get all distinct values? Thanks.

Comment (travisturner) [reply fragment.go]: Hey @brucenan we actually rejected this implementation in favor of another, more performant approach. That implementation is currently only available in Molecula, our commercial offering which uses a private version of Pilosa.

**Pull request closed without merging**